### PR TITLE
udev-builtin-path_id: SAS wide ports must have num_phys > 1

### PR DESCRIPTION
**Pull request by mwilck**
Some kernel SAS drivers (e.g. smartpqi) expose ports with num_phys = 0. udev shouldn't treat these ports as wide ports.  SAS wide ports always have num_phys > 1. See comments for sas_port_add_phy() in the kernel sources.

Sample data from a smartpqi system to illustrate the issue below. Here the phy device is attached to port 0:0, which has no end devices attached and the SAS end device (where sda is attached) is associated with SAS port 0:1, which has no associated phy device. Thus num_phys for port-0:1 is 0. This is arguably wrong, but it's how smartpqi has always set up its devices in sysfs.

```
/sys/class/sas_phy/phy-0:0 -> ../../devices/pci0000:46/0000:46:02.0/0000:47:00.0/host0/scsi_host/host0/phy-0:0/sas_phy/phy-0:0 
/sys/devices/pci0000:46/0000:46:02.0/0000:47:00.0/host0/scsi_host/host0/port-0:0/phy-0:0 -> ../phy-0:0
/sys/devices/pci0000:46/0000:46:02.0/0000:47:00.0/host0/scsi_host/host0/phy-0:0/port -> ../port-0:0

/sys/class/sas_device/end_device-0:1 -> ../../devices/pci0000:46/0000:46:02.0/0000:47:00.0/host0/scsi_host/host0/port-0:1/end_device-0:1/sas_device/end_device-0:1 
/sys/class/block/sda -> ../../devices/pci0000:46/0000:46:02.0/0000:47:00.0/host0/scsi_host/host0/port-0:1/end_device-0:1/target0:0:0/0:0:0:0/block/sda
```

### CODE REVIEW

**Comment by mwilck** [comment]
@fbuihuu, @yuwata for information

File: src/udev/udev-builtin-path_id.c [hunk 1]
```diff
@@ -24,6 +24,7 @@
 #include "sysexits.h"
 #include "udev-builtin.h"
 #include "udev-util.h"
+#include "device-private.h"
```
Comment (poettering) [inline src/udev/udev-builtin-path_id.c]: not sorted.

File: src/udev/udev-builtin-path_id.c [hunk 2]
```diff
@@ -182,11 +184,10 @@ static sd_device *handle_scsi_sas(sd_device *parent, char **path) {
         /* Get port device */
         if (sd_device_new_from_subsystem_sysname(&port_sasdev, "sas_port", sysname) < 0)
                 return NULL;
-        if (sd_device_get_sysattr_value(port_sasdev, "num_phys", &phy_count) < 0)
+        if (device_get_sysattr_unsigned(port_sasdev, "num_phys", &num_phys) < 0)
                 return NULL;
-
         /* Check if we are simple disk */
-        if (!streq(phy_count, "1"))
+        if (num_phys > 1)
```
Comment (poettering) [inline src/udev/udev-builtin-path_id.c:190]: please add a comment explaining this.

**Comment by poettering** [comment]
so this is ulimately a kernel/driver bug as I understand? why not file a bug/patch against the kernel driver in question?

**Comment by mwilck** [comment]
It's both a kernel and an udev bug. udev currently interprets "0" as "wide", which is wrong. That's what this PR is about.

We have reported that to the smartpqi maintainers too, of course. They acknowledged that the way the driver assigns its phys and devices is wrong, but they are reluctant to change that. I suppose some 3rd party software depends on this assignment but I don't know any details.

File: src/udev/udev-builtin-path_id.c [hunk 3]
```diff
@@ -159,7 +160,8 @@ static sd_device *handle_scsi_sas(sd_device *parent, char **path) {
         _cleanup_(sd_device_unrefp) sd_device *target_sasdev = NULL, *expander_sasdev = NULL, *port_sasdev = NULL;
         const char *sas_address = NULL;
         const char *phy_id;
-        const char *phy_count, *sysname;
+        const char *sysname;
+        unsigned int num_phys;
```
Comment (poettering) [inline src/udev/udev-builtin-path_id.c]: in our codebase we call the type just "unsigned", not "unsigned int"

**Pull request merged**